src/realm/object-store/c_api/realm.cpp

```diff
@@ -17,6 +17,11 @@ realm_refresh_callback_token::~realm_refresh_callback_token()
     realm::c_api::CBindingContext::get(*m_realm).realm_pending_refresh_callbacks().remove(m_token);
 }
 
+realm_thread_observer_token::~realm_thread_observer_token()
```

Comment:
I am just resetting the `g_ptr` here, but the token is actually not keeping any reference to the C_API implementation for listening to object store events. 
An alternative would be to own the implementation `ptr` and release it once the token is released as well.